src/tyr/route_serializer.cc

```diff
@@ -197,46 +197,28 @@ namespace {
       return street;
     }
 
-    // Serialize locations (called waypoints in OSRM). Waypoints are described here:
-    //     http://project-osrm.org/docs/v5.5.1/api/#waypoint-object
-    json::ArrayPtr waypoints(const std::list<valhalla::odin::TripDirections>& legs){
-      int index = 0;
-      auto waypoints = json::array({});
-      for (auto leg = legs.begin(); leg != legs.end(); ++leg) {
-        for (auto location = leg->location().begin() + index; location != leg->location().end(); ++location) {
-          index = 1;
+    // Serialize waypoints for optimized route. Note that OSRM retains the
+    // original location order, and stores an index for the waypoint index in
+    // the optimized sequence.
+    json::ArrayPtr waypoints(const google::protobuf::RepeatedPtrField<odin::Location>& locs) {
+      std::vector<std::pair<valhalla::odin::Location, uint32_t>> locations;
```

Comment:
why copy the locations at all, just store a vector of indices and then in the loop at the bottom call `waypoints->emplace_back(osrm::waypoint(loc.Get(location_index), false, true, location_index));`